spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/listener/AbstractHazelcastEventListener.java

```diff
@@ -0,0 +1,101 @@
+ * limitations under the License.
+ */
+
+package org.springframework.integration.hazelcast.listener;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.EventObject;
+import java.util.HashSet;
+import java.util.Set;
+
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.MultiMap;
+
+import org.springframework.integration.hazelcast.common.CacheListeningPolicyType;
+import org.springframework.integration.hazelcast.context.ApplicationContextStartEventHandler;
+import org.springframework.integration.hazelcast.inbound.HazelcastMessageProducer;
+import org.springframework.messaging.support.MessageBuilder;
+
+/**
+ * This is abstract Hazelcast event listener for entry/item/message events.
+ * 
+ * @author Eren Avsarogullari
+ * @since 1.0.0
+ *
+ */
+public abstract class AbstractHazelcastEventListener {
+
+	private HazelcastMessageProducer hazelcastInboundChannelAdapter;
+
+	public AbstractHazelcastEventListener(HazelcastMessageProducer hazelcastInboundChannelAdapter) {
+		this.hazelcastInboundChannelAdapter = hazelcastInboundChannelAdapter;
+	}
+
+	public HazelcastMessageProducer getHazelcastInboundChannelAdapter() {
+		return hazelcastInboundChannelAdapter;
+	}
+
+	protected abstract void processEvent(EventObject event);
+
+	protected Set<SocketAddress> getLocalSocketAddresses() {
+		Set<SocketAddress> localSocketAddressesSet = new HashSet<>();
+		for (HazelcastInstance hazelcastInstance : Hazelcast.getAllHazelcastInstances()) {
+			localSocketAddressesSet.add(hazelcastInstance.getLocalEndpoint().getSocketAddress());
+		}
+
+		return localSocketAddressesSet;
+	}
+
+	protected void sendMessage(EventObject event, InetSocketAddress socketAddress,
+			CacheListeningPolicyType cacheListeningPolicyType) {
+		
+		if (CacheListeningPolicyType.ALL == cacheListeningPolicyType
+				|| isEventAcceptable(socketAddress)) {
+
+			getHazelcastInboundChannelAdapter().sendMessage(
+					MessageBuilder.withPayload(event).build());
+
+		}
+	}
+
+	private boolean isEventAcceptable(InetSocketAddress socketAddress) {
+		Set<SocketAddress> localSocketAddressesSet = getLocalSocketAddresses();
+		if ((!localSocketAddressesSet.isEmpty())
+				&& (localSocketAddressesSet.contains(socketAddress) || 
+						isEventAcceptable(localSocketAddressesSet, socketAddress))) {
+			return true;
+		}
+
+		return false;
+	}
+
+	private boolean isEventAcceptable(Set<SocketAddress> localSocketAddressesSet,
+			InetSocketAddress socketAddressOfEvent) {
+		HazelcastInstance hazelcastInstance = Hazelcast.getAllHazelcastInstances().iterator().next();
+		MultiMap<SocketAddress, SocketAddress> configMultiMap = hazelcastInstance
+				.getMultiMap(ApplicationContextStartEventHandler.HZ_INTERNAL_CONFIGURATION_MULTI_MAP);
+		if (configMultiMap.size() > 0
+				&& !configMultiMap.values().contains(socketAddressOfEvent)
+				&& localSocketAddressesSet.contains(configMultiMap.keySet().iterator().next())) {
```

Comment:
Good. Lets take care of that from the separate JIRA: https://jira.spring.io/browse/INTEXT-151!
For the `IAtomicLong` we can follow with logic provided in the `RoundRobinLoadBalancingStrategy`.

From other side the `LoadBalancingStrategy` can be as some `DistributedObject`, which we retrieve on each event, get the index from it, e.g. random and store back to the Hazelcast, if it is necessary.

OK we can continue to discuss further in that JIRA